### PR TITLE
Checkout: Add entities for checkout fields that can be toggled on/off

### LINKED ISSUE #42773: Move Phone, Company, and Address 2 to be part of fields definitions instead of being hard-coded into UI.
Currently, the logic of showing/hiding, and requiring the Phone, Company, and Address 2 fields is saved to the Checkout attributes (and HTML), this causes a problem in which you can't access those values outside of Checkout, think My Account or Store API, this might causes conflicting expectations, as My account is still using the legacy options.

The expected outcome (without diving too deep into implementation detail) is that the value of those fields should be saved and read from outside the Checkout block, so that My Account has access to it and that Store API has access to it as well.

### PR DESCRIPTION
**Pull request by mikejolley**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Moves the visibility settings (hidden and required) for phone, company, and address 2 inputs to global settings instead of block attributes.

This allows the settings to be visible globally across all blocks as well as via the Store API. This approach has several benefits since we know at store level which fields are enabled. The main drawback is you can no longer have different settings per checkout block, but multiple checkout blocks is an edge use case.

Attributes from the checkout block are migrated to the global settings on access if the main checkout page is the checkout block, and the settings are blank. **We could look into migrating even if the settings are not blank, TBD**.

These global settings were already in place:

- `woocommerce_checkout_company_field`
- `woocommerce_checkout_address_2_field`
- `woocommerce_checkout_phone_field`

I believe these are used in the customizer (see `WC_Shop_Customizer::add_checkout_section` and contain values `hidden`, `required`, and `optional`. We match this in the API and the blocks.

In the editor, instead of the field settings being saved to attributes we use entities linked to wp settings. To handle the preview, this PR uses the existing `previewData` to populate the core fields combined with `getEditedEntityRecord` for said fields. On the frontend, this is not needed.

~There is once big caveat with this PR and that is the previously used attributes are discarded. Because they live at block level, I cannot see a good way to persist these globally. Therefore on upgrade, some fields on checkout may change behaviour until the merchant edits the block. This shouldn't be breaking for customers since we'll use the defaults (phone optional, address optional, company hidden).~

Closes #42773

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Edit the page containing the checkout block
2. Focus on the address block
3. Try toggling on/off phone/company/apartment fields, as well as the required/optional checkbox and ensure the preview output matches your selections
4. Save changes
5. Go to the frontend block checkout
6. Confirm the changes made in the editor are reflected on the checkout
7. Go to the frontend legacy shortcode checkout
8. The phone/address 2/company fields should also match the changes you made when editing the checkout block

To test the migration process (this is also covered by tests):

1. On a clean install setup the checkout block before using this PR. The checkout block must be on the main checkout page.
2. Change some field settings e.g. make phone required, hide address 2, etc. Save.
3. Use this PR.
4. Confirm the block settings were persisted.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Comment by alexflorisca** [comment]
Really cool that you got this working, I was just about to look at that the issue this solves.

> There is once big caveat with this PR and that is the previously used attributes are discarded.

Thinking out loud here, but couldn't we grab the values from the checkout block attributes using `parse_blocks` and use those as the initial values of the entities?

**Comment by mikejolley** [comment]
> Thinking out loud here, but couldn't we grab the values from the checkout block attributes using parse_blocks and use those as the initial values of the entities?

The settings exists in the customiser so whatever we do needs to work with both. Maybe we could set the defaults using existing attributes. Ill look into it.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/address-line-2-field.tsx [hunk 1]
```diff
@@ -20,12 +21,20 @@ const AddressLine2Field = < T extends AddressFormValues | ContactFormValues >( {
...
 		() => Boolean( value ) || isFieldRequired
 	);
 
+	// Re-render if the isFieldVisible prop changes.
+	useEffect( () => {
+		if ( previousIsFieldRequired !== isFieldRequired ) {
+			setIsFieldVisible( Boolean( value ) || isFieldRequired );
+		}
+	}, [ value, previousIsFieldRequired, isFieldRequired ] );
+
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/address-line-2-field.tsx:37]: Why is this needed?

**Comment by alexflorisca** [comment]
I've added myself as a reviewer because I'm curious how this works. Won't get to it until Friday or next week though so feel free to go with Nadirs review

File: plugins/woocommerce-blocks/assets/js/blocks/checkout/inner-blocks/checkout-billing-address-block/block.tsx [hunk 2]
```diff
@@ -1,41 +1,31 @@
...
-	showApartmentField: boolean;
-	requireApartmentField: boolean;
-	showPhoneField: boolean;
-	requirePhoneField: boolean;
-} ): JSX.Element => {
-	const { billingAddress, setShippingAddress, useBillingAsShipping } =
-		useCheckoutAddress();
+const Block = (): JSX.Element => {
+	const {
+		defaultFields,
```
Comment (alexflorisca) [inline plugins/woocommerce-blocks/assets/js/blocks/checkout/inner-blocks/checkout-billing-address-block/block.tsx:24]: I wonder if we can rename this to something that convey's what it is a bit better. Maybe `addressFieldsConfig`? or something like that

File: plugins/woocommerce/src/Blocks/Utils/CartCheckoutUtils.php [hunk 3]
```diff
@@ -44,6 +44,112 @@ public static function is_checkout_block_default() {
...
+			return;
+		}
+
+		$checkout_block = $checkout_blocks[0];
+
+		if ( ! isset( $checkout_block['attrs'] ) ) {
+			return;
+		}
+
+		if ( isset( $checkout_block['attrs']['showPhoneField'], $checkout_block['attrs']['requirePhoneField'] ) ) {
```
Comment (alexflorisca) [inline plugins/woocommerce/src/Blocks/Utils/CartCheckoutUtils.php]: Not sure this logic is quite right. The attributes are only stored in markup if they are different from the defaults. So if you have a false value for either of them, it just means that it takes the default, not that it's necessarily hidden.

**Review by alexflorisca — CHANGES_REQUESTED** [review]
Nice work Mike, chunky PR but it tests pretty well and learned quite a lot from it.

I think there's a problem with the migration logic and left a couple of other comments. Another thing I noticed is that while you changed the name of the editor control from Apartment -> Address Line 2, it's still registered as Apartment [here](https://github.com/woocommerce/woocommerce/blob/add/checkout-field-entities-42773/plugins/woocommerce/src/Blocks/BlockTypes/Checkout.php#L105). This shows up when you save so it would be good to change that as well. Not sure if there are other implications to this?

Screenshot 2024-11-25 at 15 20 09

File: plugins/woocommerce/src/Blocks/Utils/CartCheckoutUtils.php (hunk 3 above)
Comment (mikejolley) [reply plugins/woocommerce/src/Blocks/Utils/CartCheckoutUtils.php]: Are you suggesting we migrate regardless? I was worried we'd be changing site settings if the user hadn't changed the default options. This could impact the shortcodes too. 

I think its only safe to migrate settings if we're sure the merchant has changed these settings, so I would be inclined to change this to:

```
if ( $checkout_block['attrs']['showPhoneField'] ) {
	update_option( 'woocommerce_checkout_phone_field', $checkout_block['attrs']['requirePhoneField'] ? 'required' : 'optional' );
}
```

And leave current defaults alone otherwise. It would mean some stores see a change in fields displayed on the block checkout however.

cc @senadir

File: plugins/woocommerce-blocks/assets/js/blocks/checkout/inner-blocks/checkout-billing-address-block/block.tsx (hunk 2 above)
Comment (mikejolley) [reply plugins/woocommerce-blocks/assets/js/blocks/checkout/inner-blocks/checkout-billing-address-block/block.tsx:24]: `defaultFields` was pre-existing here: https://github.com/woocommerce/woocommerce/blob/7b8a65c640bae57a60cf250da1eb5357b9e4d9d8/plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php#L111

I think I'd rather leave it than impact other code consuming this data. At least in the context of the billing address block, `defaultFields` has some meaning.

**Comment by mikejolley** [comment]
I've renamed the field in this commit: https://github.com/woocommerce/woocommerce/pull/52784/commits/4107d79743391596af6e23d0ea60b7ff1f51dee0

File: plugins/woocommerce-blocks/assets/js/blocks/checkout/inner-blocks/checkout-billing-address-block/block.tsx (hunk 2 above)
Comment (senadir) [reply plugins/woocommerce-blocks/assets/js/blocks/checkout/inner-blocks/checkout-billing-address-block/block.tsx:24]: yeah let's leave it as is

File: plugins/woocommerce-blocks/assets/js/blocks/checkout/edit.tsx [hunk 4]
```diff
@@ -69,13 +65,93 @@ export const Edit = ( {
...
+			setLastAddress2Required( defaultFields.address_2.required );
+		}
+	}, [
+		defaultFields.phone.hidden,
+		defaultFields.company.hidden,
+		defaultFields.address_2.hidden,
+		defaultFields.phone.required,
+		defaultFields.company.required,
+		defaultFields.address_2.required,
+	] );
```
Comment (senadir) [inline plugins/woocommerce-blocks/assets/js/blocks/checkout/edit.tsx:130]: Doesn't seem scalable as we want to add those control to more fields later on but we solve it once we reach it

File: plugins/woocommerce/src/Blocks/BlockTypes/Checkout.php [hunk 5]
```diff
@@ -57,6 +58,63 @@ public function dequeue_woocommerce_core_scripts() {
 		wp_dequeue_style( 'select2' );
 	}
 
+	/**
+	 * Exposes settings exposed by the checkout block.
+	 */
+	public function register_settings() {
```
Comment (senadir) [inline plugins/woocommerce/src/Blocks/BlockTypes/Checkout.php:64]: Would be nice to limit this to only admin requests of REST as I don't (yet) think this should be a global option.

**Review by senadir — APPROVED** [review]
Love this! we might want to expand a bit on testing steps so they include upgrading from an existing site.

I tested both on an existing site and a new one and both worked well

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/address-line-2-field.tsx (hunk 1 above)
Comment (mikejolley) [reply plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/form/address-line-2-field.tsx:37]: There was some similar logic previously. This forces a re-render if you toggle the visibility of the field. Without this the changes are not reflected in the editor.

File: plugins/woocommerce-blocks/assets/js/blocks/checkout/edit.tsx (hunk 4 above)
Comment (mikejolley) [reply plugins/woocommerce-blocks/assets/js/blocks/checkout/edit.tsx:130]: Yeah I had to do this because of the way the UI works; without tracking state you would lose the previous checked state when toggling visibility. Thats because the visibility and required status is tracked within a single entity as a string.

File: plugins/woocommerce/src/Blocks/Utils/CartCheckoutUtils.php (hunk 3 above)
Comment (mikejolley) [reply plugins/woocommerce/src/Blocks/Utils/CartCheckoutUtils.php]: I've updated the logic to migrate when the attributes are missing. We assume defaults if they are not present in the attributes. See https://github.com/woocommerce/woocommerce/pull/52784/commits/05f411bfc84ec8d31a8e9c11e342325d119b1d14

**Comment by mikejolley** [comment]
@alexflorisca changed the logic you requested. Can you check again? Tests are passing.

File: plugins/woocommerce/src/Blocks/Utils/CartCheckoutUtils.php (hunk 3 above)
Comment (alexflorisca) [reply plugins/woocommerce/src/Blocks/Utils/CartCheckoutUtils.php]: > I've updated the logic to migrate when the attributes are missing. We assume defaults if they are not present in the attributes. See https://github.com/woocommerce/woocommerce/commit/05f411bfc84ec8d31a8e9c11e342325d119b1d14

Yep nice, this is what I meant

**Review by alexflorisca — APPROVED** [review]
Awesome, thanks for making those changes!

**Comment by mikejolley** [comment]
Ill rebase conflicts...

**Comment by chad1008** [comment]
@mikejolley I had an issue today where the visibility/required settings in the Customizer were actively overriding those set in the block settings.

I was able to reproduce the customer's issue on my testing site with Storefront. It's very similar to [a previously reported issue that was closed in anticipation of this PR](https://github.com/woocommerce/woocommerce/issues/52724).

Reproduction:
1. Install and activate Storefront
2. Create two pages, name them Checkout (block) and Checkout (shortcode). The names don't actually matter, this is just to keep the testing clear.
3. In the two pages, add the respective checkout block or shortcode, so we have one page using each checkout option.
4. In Checkout (block) focus on the address info block and confirm that Company Name is `Optional`
5. In WooCommerce > Settings > Advanced, select Checkout (shortcode) as the checkout page
6. In Appearance > Customize > WooCommerce > Checkout, set the Company Name field to `Required`
7. In WooCommerce > Settings > Advanced, select Checkout (block) as the checkout page
8. Open an incognito window and try to make a purchase
9. Company name should (if your site behaves like mine and this customer's) act like it's required and pop up a notice, contrary to the block settings (and we should be using the Checkout (block) page here.)

On the site that reported this issue today, I had to manually navigate to Customizer to find this setting... their theme doesn't even include a link to it so it would have been pretty much impossible to find without [support threads like this one](https://wordpress.org/support/topic/company-name-is-required-is-displayed-when-company-is-optional/#post-18116756), which is the only way I managed to solve it.

On my test site with Storefront, even once I'd deleted the page with the shortcode the Customizer settings persisted. Accessing them requires the cart page to be using the short code but once they're set they appear to stick until modified in the Customizer again, regardless of what happens with the individual block settings.

Does this feel like an unexpected side effect? Or maybe a regression since your merge?

**Comment by mikejolley** [comment]
@chad1008 its an existing issue if you cannot access the settings. Before this PR merged there was a similar case where if you used the customizer, but then lost access to it (using a block theme), shortcode cart and checkout had no way to edit these settings. The settings themselves have existing in core a long time.

After this PR you can at least edit that global setting again if using blocks because they all share the same system.

**Comment by chad1008** [comment]
> After this PR you can at least edit that global setting again if using blocks because they all share the same system.

This part is the reason I pinged... _I can't edit those global settings when using blocks for the checkout page_. I can flip the toggles, but they don't do anything on my site.

To make sure fields like `Company Name` are actually optional, I have to make sure Customizer says they are. The block setting appears to get ignored. I just tried it again on my test site with the same result:

Even if the block setting says `Optional`, I can't place an order because `Company Name` is required. I have to get into the Customizer (which means creating a shortcode-powered checkout page and selecting it in WC settings) and make sure `Company Name` is `Optional` in there (and then put my block-powered checkout back in place).

Maybe this is only a problem for sites that had a shortcode-based checkout before this PR, and then switched to a block-based checkout after this PR merged?

**Comment by mikejolley** [comment]
@chad1008 can you clarify which version of WC you're using? This PR has been merged but its not due to ship until WC 9.6 so theres no way you're using it unless you pulled trunk from GitHub.

This PR syncs those settings. So if you make a change via customizer, that same change will apply to blocks and vice versa.

**Comment by chad1008** [comment]
Ah, that's the missing piece, my bad...

PHgBLMeqy8

I wasn't thinking of the release cycle and that "merged" doesn't mean it's actually out in the wild yet. Sorry for the confusion and thanks for clarifying... glad it's not a regression or something along those lines.

**Review by Jinksi — COMMENTED** [review]
I've conducted additional testing of this PR as part of the _WooCommerce 9.6 – Release testing by Contributors_ on behalf of @woocommerce/helix.

It is working well on a new Pressable site – checkout field options modified via the checkout block editor are persisted across the checkout block, legacy shortcode checkout and my account shipping/billing address forms.

Kudos @mikejolley for the clear testing instructions, which is especially helpful for those of us who don't often work with checkout.

One concern I noted that may already be a known tradeoff: on upgrading from a WC 9.5.1 site that has both a legacy shortcode and blocks checkout (presumably a weird edge case), field options were inherited from the **legacy checkout shortcode** field options and not the **checkout block** field options.

**Testing steps followed:**

- [x] Pressable site created 
- [x] Onboarding wizard
  - [x] Installed all free features
  - [x] Connect to WordPress.com when prompted
- [x] Confirm top plugins installed
  - [x] Install and activate top plugins
    ```
      wp plugin install --activate akismet jetpack gutenberg crowdsignal-forms jetpack-protect jetpack-boost elementor contact-form-7 wpforms-lite google-site-kit google-analytics-for-wordpress google-listings-and-ads woocommerce-payments woocommerce-services insert-headers-and-footers mailpoet page-optimize wordpress-seo
    ```
- [x] Generated products with [wc-smooth-generator](https://github.com/woocommerce/wc-smooth-generator)
- [x] Install current WC version 9.5.1
- [x] Enabled WooPayments in Sandbox mode to allow checkout payment processing
- [x] Modified checkout block field options ✅
- [x] Created a legacy shortcode checkout page – field options modified above are not applied
- [x] Upgrade to WC 9.6 pre-release version https://github.com/woocommerce/woocommerce/releases/tag/9.6.0-beta.1
- ⚠️ **Note** Field options were inherited from legacy checkout shortcode options – checkout block field options were not persisted
- [x] Upon modifying the block checkout field settings, ensured the checkout field settings were applied to:
  - [x] Blocks checkout frontend ✅
  - [x] Legacy checkout frontend ✅
  - [x] My account shipping/billing address forms ✅
- [x] Modified block checkout field settings – refreshed edit page to ensure settings persisted ✅
- [x] Ensured that new settings were applied to:
  - [x] Blocks checkout frontend ✅
  - [x] Legacy checkout frontend ✅
  - [x] My account shipping/billing address forms ✅
- [x] Confirmed legacy checkout shortcode order completed successfully ✅
- [x] Out of curiosity, I downgraded to WC 9.5.1. The checkout field options continued to be applied to:
  - [x] Blocks checkout frontend ✅
  - [x] Legacy checkout frontend ✅
  - [x] My account shipping/billing address forms ✅

**Comment by Jinksi** [comment]
> One concern I noted that may already be a known tradeoff: on upgrading from a WC 9.5.1 site that has both a legacy shortcode and blocks checkout (presumably a weird edge case), field options were inherited from the **legacy checkout shortcode** field options and not the **checkout block** field options.

I've opened #54137 to track this issue.
